pkg/gadgets/trace/dns/tracer/gc.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !withoutebpf
+
+package tracer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cilium/ebpf"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+// Delay between each garbage collection run.
+const garbageCollectorInterval = 1 * time.Second
+
+// startGarbageCollector runs a background goroutine to delete old query timestamps
+// from the DNS query_map. This ensures that queries that never receive a response
+// are deleted from the map.
+//
+// The garbage collector goroutine terminates when the context is done.
+func startGarbageCollector(ctx context.Context, logger logger.Logger, gadgetParams *params.Params, queryMap *ebpf.Map) error {
+	dnsTimeout := gadgetParams.Get(ParamDNSTimeout).AsDuration()
+	if dnsTimeout <= 0 {
+		return fmt.Errorf("DNS timeout must be > 0")
+	}
+
+	logger.Debugf("starting garbage collection for DNS tracer with dnsTimeout %s", dnsTimeout)
+	go func() {
+		ticker := time.NewTicker(garbageCollectorInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Debugf("stopping garbage collection for DNS tracer")
+				return
+
+			case <-ticker.C:
+				logger.Debugf("executing DNS query map garbage collection")
+				collectGarbage(logger, dnsTimeout, queryMap)
+			}
+		}
+	}()
+
+	return nil
+}
+
+func collectGarbage(logger logger.Logger, dnsTimeout time.Duration, queryMap *ebpf.Map) {
```

Comment:
Nice, I didn't realize these were available. Updated to use `BatchLookup` and `BatchDelete`. I set the batch size to 256 so worst-case this would require 1024 / 256 = 4 lookup and delete calls, but could adjust that higher or lower. Tested it by running 5000 nslookup calls to a bad IP in parallel and verified from the logs that 1024 entries were collected.